### PR TITLE
Implement MCCP2 compression support

### PR DESCRIPTION
**Pull request by Tiger-Tom**
Implements MCCP2 compression support (https://mudhalla.net/tintin/protocols/mccp/), resolving https://github.com/MUME/play-mume/issues/3

### CODE REVIEW

File: src/js/decafmud.socket.websocket.js [hunk 1]
```diff
@@ -199,6 +199,9 @@ DecafWebSocket.prototype.onMessage = function(websocket, event) {
 	var reader = new FileReader();
 	reader.onload = function(e) {
 		var u8array = new Uint8Array(e.target.result);
+		this.decaf.socketData(u8array);
+		return;
```
Comment (nschimme) [inline src/js/decafmud.socket.websocket.js]: We should just delete the dead code following the return statement as it is never run

File: src/js/decafmud.js [hunk 2]
```diff
@@ -1598,17 +1641,25 @@ DecafMUD.prototype.encode = function(data) {
 /** Read through data, only stopping for TELNET sequences. Pass data through
  *  towards the display handler. */
 DecafMUD.prototype.processBuffer = function() {
-	var enc, data, ind, out;
-
-	data = this.inbuf.join(''), IAC = DecafMUD.TN.IAC, left='';
+	var enc, data, ind, out, compressBuf, temp;
+	data = [];
+	// Looks nicer than this.inbuf.map(s=>Array.from(s)).flat().map(s=>String.fromCharCode(s)).join('');
```
Comment (nschimme) [inline src/js/decafmud.js]: No need to leave a comment with why this looks nicer. I would instead explain why it's sometimes an array or a string

File: src/js/decafmud.js [hunk 3]
```diff
@@ -1532,6 +1565,16 @@ DecafMUD.prototype.socketClosed = function() {
 /** Called by the socket when data arrives. */
 DecafMUD.prototype.socketData = function(data) {
 	// Push the text onto the inbuf.
+	if (this.compressingv2) { // Decompress it first!
+		try {
+			data = this.decompressStream.decompress(data);
+		} catch (e) {
+			console.error('MCCP2 compression disabled because ' + e);
```
Comment (nschimme) [inline src/js/decafmud.js]: Should we instead use this.decaf.debugString() or this.decaf.error() ? The later will show an error to the end-user which might not be very useful.

File: src/js/decafmud.js [hunk 4]
```diff
@@ -1872,7 +1940,7 @@ DecafMUD.options = {
 		handlecolor	: true,
 		fgclass		: 'c',
 		bgclass		: 'b',
-        fntclass    : 'fnt',
+		fntclass    : 'fnt',
```
Comment (nschimme) [inline src/js/decafmud.js]: This change doesn't seem to be relevant to this PR.

**Review by nschimme — CHANGES_REQUESTED** [review]

File: src/js/decafmud.socket.websocket.js (hunk 1 above)
Comment (Tiger-Tom) [reply src/js/decafmud.socket.websocket.js]: Fixed, thanks! https://github.com/Tiger-Tom/DecafMUD/commit/09a046b77738f7d16d0bece69d55f47bc70bdf23

File: src/js/decafmud.js (hunk 2 above)
Comment (Tiger-Tom) [reply src/js/decafmud.js]: https://github.com/Tiger-Tom/DecafMUD/commit/aa44499e6c0fc0d9a8b96a250f20dc26930bee3d

File: src/js/decafmud.js (hunk 4 above)
Comment (Tiger-Tom) [reply src/js/decafmud.js]: https://github.com/Tiger-Tom/DecafMUD/commit/aa44499e6c0fc0d9a8b96a250f20dc26930bee3d
I think that this was just a result of the flattening of leading spaces to tabs.

File: src/js/decafmud.js (hunk 3 above)
Comment (Tiger-Tom) [reply src/js/decafmud.js]: https://github.com/Tiger-Tom/DecafMUD/commit/25922361baa093d3d776952f18616eba68a0ad20
